### ISSUE TITLE
ack 'foo\w+' -w -i is surprisingly slow

### ISSUE THREAD

**Issue by petdance** [beyondgrep/ack3#181]
```
[22:29:28] (dev) clifford:~/ack3 $ time ack 'foo\w+' -w ~/parrot/ > /dev/null

real	0m0.396s
user	0m0.329s
sys	0m0.066s
[22:29:33] (dev) clifford:~/ack3 $ time ack 'foo\w+' -S ~/parrot/ > /dev/null

real	0m0.440s
user	0m0.362s
sys	0m0.078s
[22:29:38] (dev) clifford:~/ack3 $ time ack 'foo\w+' -w -S ~/parrot/ > /dev/null

real	0m2.954s
user	0m2.868s
sys	0m0.086s
```

What is it about the combination of `\w+` and `-w` and `-S` that makes things slow?

We need to make sure that the results of that search are correct, too.

**Comment by petdance** [comment]
I added some diagnostics to see what the built regex looks like.  Perhaps it's too complex?

    ./ack-standalone 'foo\w+' -S -w foo

builds this regex

    (?^m:(?i)(?:^|\b|\s)\K(?:foo\w+)(?=\s|\b|$))

**Comment by petdance** [comment]
It's `-i` that is the problem.  `-S` is just a variant.

**Comment by petdance** [comment]
The slow down is even worse under Perl 5.10.1, the oldest Perl we support.

**Comment by petdance** [comment]
This is fixed at e7aa7aa.  It will be released in 3.1.1 or 3.2.0.
